### PR TITLE
OZ-120: Improved Docker version check in start demo script.

### PR DESCRIPTION
**Pull request by kdaud**
Refer to [OZ-49](https://mekomsolutions.atlassian.net/browse/OZ-49)

Responsible ticket: https://mekomsolutions.atlassian.net/browse/OZ-120

Additional conversation via [Slack](https://mekomsolutions.slack.com/archives/G421UNF5L/p1680593461771949) !

[OZ-49]: https://mekomsolutions.atlassian.net/browse/OZ-49?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: start-demo.sh [hunk 1]
```diff
@@ -25,14 +25,7 @@ DOCKER_SERVER_VERSION_MAJOR=$(echo "$DOCKER_SERVER_VERSION"| cut -d'.' -f 1)
...
-if [ "${DOCKER_SERVER_VERSION_MAJOR}" -ge 20 ] && \
-   [ "${DOCKER_SERVER_VERSION_MINOR}" -ge 10 ]  && \
-   [ "${DOCKER_SERVER_VERSION_BUILD}" -ge 13 ]; then
-    echo "Docker version >= 20.10.13, using Docker Compose V2"
-    if command -v gp version &> /dev/null; then
-        export PROXY_TLS="-DgitPodEnvironment"    
-    fi
-    docker compose up -d --build
-else
-    echo "Docker versions < 20.10.13 are not supported"
```
Comment (mks-d) [inline start-demo.sh:37]: @kdaud but why remove the check on Docker versions?

Comment (kdaud) [reply start-demo.sh:37]: @enyachoke could you throw a comment here?

File: start-demo.sh [hunk 2]
```diff
@@ -25,14 +25,7 @@ DOCKER_SERVER_VERSION_MAJOR=$(echo "$DOCKER_SERVER_VERSION"| cut -d'.' -f 1)
...
-    if command -v gp version &> /dev/null; then
-        export PROXY_TLS="-DgitPodEnvironment"    
-    fi
-    docker compose up -d --build
-else
-    echo "Docker versions < 20.10.13 are not supported" 
+if command -v gp version &> /dev/null; then
+  export PROXY_TLS="-DgitPodEnvironment"
 fi
+docker compose up -d --build
```
Comment (Ruhanga) [inline start-demo.sh:31]:
```suggestion
INSTALLED_DOCKER_VERSION=$(docker version -f "{{.Server.Version}}")
MINIMUM_REQUIRED_DOCKER_VERSION="^((([2-9][1-9]|[3-9][0]|[0-9]{3,}).*)|(20\.([0-9]{3,}|[1-9][1-9]|[2-9][0]).*)|(20\.10\.([0-9]{3,}|[2-9][0-9]|[1][3-9])))"

if [[ $INSTALLED_DOCKER_VERSION =~ $MINIMUM_REQUIRED_DOCKER_VERSION ]]; then
    echo "Docker version >= 20.10.13, using Docker Compose V2"
    if command -v gp version &> /dev/null; then
        export PROXY_TLS="-DgitPodEnvironment"    
    fi
    docker compose up -d --build
else
    echo "Docker versions < 20.10.13 are not supported" 
fi
```

**Review by Ruhanga — CHANGES_REQUESTED** [review]
Thanks for being pro-active on this @kdaud. It's not okay to eliminate the docker version check. I'd suppose doing something along the following lines below.

File: start-demo.sh (hunk 1 above)
Comment (kdaud) [reply start-demo.sh]: > @kdaud but why remove the check on Docker versions?

Removing the check seemed to resolve the issue with gitpod however, we indeed need the logic. So @Ruhanga suggested to make the check robust which is working well on gitpod environment.

**Comment by kdaud** [comment]
@Ruhanga have tested out the changes on gitpod and the check is working fine, Ozone instances are able to open up well on their respective ports.

**Review by Ruhanga — CHANGES_REQUESTED** [review]
Thanks @kdaud. One last one like I said in my comment above. Can you also add the same if check in the `destroy-demo.sh` script.

**Comment by kdaud** [comment]
> Can you also add the same if check logic in the `destroy-demo.sh` script?

Done!

File: destroy-demo.sh [hunk 3]
```diff
@@ -19,12 +19,15 @@ DOCKER_SERVER_VERSION_MAJOR=$(echo "$DOCKER_SERVER_VERSION"| cut -d'.' -f 1)
 DOCKER_SERVER_VERSION_MINOR=$(echo "$DOCKER_SERVER_VERSION"| cut -d'.' -f 2)
 DOCKER_SERVER_VERSION_BUILD=$(echo "$DOCKER_SERVER_VERSION"| cut -d'.' -f 3)
```
Comment (Ruhanga) [inline destroy-demo.sh:20]: Could you remove these unused variables, `DOCKER_SERVER_VERSION_MAJOR`, `DOCKER_SERVER_VERSION_MINOR`, `DOCKER_SERVER_VERSION_BUILD`...

File: destroy-demo.sh [hunk 4]
```diff
@@ -19,12 +19,15 @@ DOCKER_SERVER_VERSION_MAJOR=$(echo "$DOCKER_SERVER_VERSION"| cut -d'.' -f 1)
...
+if [[ $INSTALLED_DOCKER_VERSION =~ $MINIMUM_REQUIRED_DOCKER_VERSION_REGEX ]]; then
+    echo "Docker version >= 20.10.13, using Docker Compose V2"
+    if command -v gp version &> /dev/null; then
+        export PROXY_TLS="-DgitPodEnvironment"
+    fi
     docker compose down -v
 else
-    echo "Docker version < 20.10.13 using docker compose v1"
+    echo "Docker versions < 20.10.13 are not supported"
     docker-compose down -v
```
Comment (Ruhanga) [inline destroy-demo.sh:32]:
```suggestion
```

File: destroy-demo.sh [hunk 5]
```diff
@@ -19,12 +19,15 @@ DOCKER_SERVER_VERSION_MAJOR=$(echo "$DOCKER_SERVER_VERSION"| cut -d'.' -f 1)
...
-   [ "${DOCKER_SERVER_VERSION_MINOR}" -ge 10 ]  && \
-   [ "${DOCKER_SERVER_VERSION_BUILD}" -ge 13 ]; then
-    echo "Docker version >= 20.10.13 using docker compose v2"
+INSTALLED_DOCKER_VERSION=$(docker version -f "{{.Server.Version}}")
+MINIMUM_REQUIRED_DOCKER_VERSION_REGEX="^((([2-9][1-9]|[3-9][0]|[0-9]{3,}).*)|(20\.([0-9]{3,}|[1-9][1-9]|[2-9][0]).*)|(20\.10\.([0-9]{3,}|[2-9][0-9]|[1][3-9])))"
+if [[ $INSTALLED_DOCKER_VERSION =~ $MINIMUM_REQUIRED_DOCKER_VERSION_REGEX ]]; then
+    echo "Docker version >= 20.10.13, using Docker Compose V2"
+    if command -v gp version &> /dev/null; then
+        export PROXY_TLS="-DgitPodEnvironment"
+    fi
```
Comment (Ruhanga) [inline destroy-demo.sh:28]:
```suggestion
```

**Review by Ruhanga — CHANGES_REQUESTED** [review]
Generally looks good. I've a few  final comments and suggestions after we'll be good to go.

**Comment by kdaud** [comment]
@Ruhanga have addressed the reviews and hopefully we are good to go.

File: destroy-demo.sh (hunk 5 above)
Comment (Ruhanga) [reply destroy-demo.sh:24]: @kdaud did you miss this?

Comment (kdaud) [reply destroy-demo.sh]: Sorry I had missed it out but have operated it.

**Review by Ruhanga — APPROVED** [review]
Thanks @kdaud. LGTM if you've double confirmed this is woking correctly.

**Pull request merged**